File: journalist-client/src/commands/vaults.ts
import { OpenVaultOutcome } from "../model/bindings/OpenVaultOutcome";
import { VaultState } from "../model/bindings/VaultState";
import { useErrorStore } from "../state/errors";
import {
  invokeWithErrorMessage,
  invokeWithSilencedErrorMessage,
} from "./invokeWithErrorMessage";

export const getVaultState = (): Promise<VaultState | null> => {
  return invokeWithErrorMessage("get_vault_state");
};

export const unlockVault = async (
  profile: string,
  path: string,
  password: string,
): Promise<void> => {
  const outcome = await invokeWithErrorMessage<OpenVaultOutcome>(
    "unlock_vault",
    { profile, path, password },
  );

  if (outcome.type === "openedOffline") {
    useErrorStore
      .getState()
      .addWarning(
        "Vault opened while offline, cannot perform preflight checks",
      );
  }

  if (outcome.type === "openedOnline") {
    for (const missingInVault of outcome.orgPksMissingInVault) {
      useErrorStore
        .getState()
        .addWarning(
          `Organization public key "${missingInVault.substring(0, 8)}" found in API but not in vault. This can cause issues, talk to the admin team.`,
        );
    }

    for (const missingInApi of outcome.orgPksMissingInApi) {
      useErrorStore
        .getState()
        .addWarning(
          `Trust anchor ${missingInApi.substring(0, 8)} found in vault but not in API.`,
        );
    }
  }
};

export const softLockVault = (): Promise<VaultState | null> => {
  return invokeWithErrorMessage("soft_lock_vault");
};

export const unlockSoftLockedVault = (
  password: string,
): Promise<VaultState | null> => {
  return invokeWithErrorMessage("unlock_soft_locked_vault", { password });
};

export const getColocatedPassword = (path: string): Promise<string | null> => {
  return invokeWithSilencedErrorMessage("get_colocated_password", { path });
};

export const addTrustAnchor = (path: string): Promise<string | null> => {
  return invokeWithErrorMessage("add_trust_anchor", { path });
};

export const sendDesktopNotification = (args: {
  title?: string;
  body: string;
}): Promise<void> => {
  return invokeWithErrorMessage("send_notification", args);
};

File: journalist-client/src-tauri/src/commands/vaults.rs
use common::{
    clap::Stage,
    crypto::keys::{public_key::PublicKey, untrusted::signing::UntrustedSignedPublicSigningKey},
    protocol::keys::UntrustedOrganizationPublicKey,
    protocol::{keys::AnchorOrganizationPublicKey, roles::Organization},
};
use snafu::{OptionExt as _, ResultExt as _};
use std::{collections::HashSet, path::Path};
use tauri::State;

use crate::{
    app_state::AppStateHandle,
    error::{CommandError, GenericSnafu, MissingProfileSnafu, VaultSnafu},
    model::{OpenVaultOutcome, Profiles, VaultState},
};

#[tauri::command]
pub async fn get_vault_state(
    app: State<'_, AppStateHandle>,
) -> Result<Option<VaultState>, CommandError> {
    app.inner().vault_state().await.context(VaultSnafu {
        failed_to: "get vault state",
    })
}

#[tauri::command]
pub async fn unlock_vault(
    path: &Path,
    stage: String,
    password: &str,
    app: State<'_, AppStateHandle>,
    profiles: State<'_, Profiles>,
) -> Result<OpenVaultOutcome, CommandError> {
    let profiles = profiles.inner();

    let api_url = profiles.api_url(&stage).context(MissingProfileSnafu)?;

    let stage = Stage::from_guardian_str(stage.as_str())
        .ok()
        .context(GenericSnafu {
            ctx: "No trust anchors exist for stage provide",
        })?;

    let (vault, api_client) = app
        .inner()
        .unlock_vault(stage, api_url, path, password)
        .await
        .context(VaultSnafu {
            failed_to: "unlock vault, is your password correct?",
        })?;

    let Ok(keys) = api_client.get_public_keys().await else {
        // If we're not able to get the public key hierarchy we can't
        // compare our org keys to see if theres a mismatch.
        // We perhaps ought to queue up something to run this process
        // when the internet comes back.
        return Ok(OpenVaultOutcome::OpenedOffline);
    };

    let org_pks: HashSet<UntrustedSignedPublicSigningKey<Organization>> = vault
        .org_pks()
        .context(VaultSnafu {
            failed_to: "get trust anchors for profile",
        })?
        .into_iter()
        .map(|org_pk: AnchorOrganizationPublicKey| org_pk.to_non_anchor().to_untrusted())
        .collect::<HashSet<_>>();

    let api_org_pks = keys
        .keys
        .org_pk_iter()
        .cloned()
        .collect::<HashSet<UntrustedOrganizationPublicKey>>();

    let org_pks_missing_in_api = org_pks
        .difference(&api_org_pks)
        .map(|org_pk| org_pk.public_key_hex())
        .collect();

    // TODO if the API has org keys we don't have, it means that Sentinel needs to be updated,
    // or the user has the wrong stage selected. We should lock the vault and inform the user
    // rather than proceeding with a warning.
    // https://github.com/guardian/coverdrop-internal/issues/3785
    let org_pks_missing_in_vault = api_org_pks
        .difference(&org_pks)
        .map(|org_pk| org_pk.public_key_hex())
        .collect();

    Ok(OpenVaultOutcome::OpenedOnline {
        org_pks_missing_in_vault,
        org_pks_missing_in_api,
    })
}

#[tauri::command]
pub async fn soft_lock_vault(
    app: State<'_, AppStateHandle>,
) -> Result<Option<VaultState>, CommandError> {
    app.inner().soft_lock_vault().await.context(VaultSnafu {
        failed_to: "soft lock vault",
    })
}

#[tauri::command]
pub async fn unlock_soft_locked_vault(
    app: State<'_, AppStateHandle>,
    password: String,
) -> Result<Option<VaultState>, CommandError> {
    app.inner()
        .unlock_soft_locked_vault(&password)
        .await
        .context(VaultSnafu {
            failed_to: "unlock soft locked vault",
        })
}

#[tauri::command]
pub async fn get_colocated_password(path: &Path) -> Result<Option<String>, CommandError> {
    let password_path = path.with_extension("password");

    let password_result = std::fs::read_to_string(&password_path);

    match password_result {
        Ok(password) => {
            if password.contains('\n') {
                return Err(GenericSnafu {
                    ctx: "Password file found, but it was not a single line",
                }
                .build());
            }

            Ok(Some(password))
        }
        Err(_) => Ok(None),
    }
}

#[tauri::command]
pub async fn send_notification(
    app: State<'_, AppStateHandle>,
    title: Option<String>,
    body: String,
) -> Result<(), CommandError> {
    let _ = app.notifications.send(title, body).await;
    Ok(())
}
